c/meterpreter/source/extensions/stdapi/server/sys/config/config.c

```diff
@@ -274,6 +274,39 @@ DWORD request_sys_config_drop_token(Remote* pRemote, Packet* pPacket)
 	return dwResult;
 }
 
+/*
+ * @brief Updates an existing thread token.
+ * @param pRemote Pointer to the \c Remote instance.
+ * @param pRequest Pointer to the \c Request packet.
+ * @returns Indication of success or failure.
+ */
+DWORD request_sys_config_update_token(Remote* pRemote, Packet* pPacket)
+{
+	Packet* pResponse = met_api->packet.create_response(pPacket);
+	DWORD dwResult = ERROR_SUCCESS;
+	HANDLE hToken = NULL;
+
+	// Get token handle from the client
+	hToken = (HANDLE)met_api->packet.get_tlv_value_qword(pPacket, TLV_TYPE_HANDLE);
+
+	// Impersonate token in the current thread
+	if (!ImpersonateLoggedOnUser(hToken))
+	{
+		dwResult = GetLastError();
+		dprintf("[UPDATE-TOKEN] Failed to impersonate token (%u)", dwResult);
+		met_api->packet.transmit_response(dwResult, pRemote, pResponse);
+		return dwResult;
+	}
+
+	// Store the token handle for future tasks
+	met_api->thread.update_token(pRemote, hToken);

```

Comment:
Let's update this to return the old token value so it can be tracked by Metasploit. This would allow Metasploit to either close the old value (which is what your module should be doing) or restore it at a later time. We want to ensure there are no handle leaks.

You can test this by using System Informer (or Process Hacker which has been replaced by System Informer). There will be a tab when inspecting the process to see what handles are open and their types. Running the `make_token` module over and over again should ensure that the handle count does not continue to climb.

Reply:
There should be no leaked handles afaik. `update_token` (i.e. `core_update_thread_token`) closes the old handle - see lines 53-56 below:

https://github.com/rapid7/metasploit-payloads/blob/2aadbeec334027e7af445a6070dc24360da6d176/c/meterpreter/source/metsrv/core.c#L34-L61

I guess If we want to return the old value to MSF to keep track of it, we would need to modify `core_update_thread_token` and not close the handle there